File: typescript/packages/mechanisms/multiversx/test/scheme.test.ts
import { describe, it, expect, vi } from 'vitest'
import { ExactMultiversXScheme } from '../src/exact/client/scheme'
import { MultiversXSigner } from '../src/signer'
import { PaymentRequirements } from '@x402/core/types'
import { Address } from '@multiversx/sdk-core'
import { ExactMultiversXPayload } from '../src/types'

const alice = new Address(Buffer.alloc(32, 1)).bech32()
const bob = new Address(Buffer.alloc(32, 2)).bech32()
const relayer = new Address(Buffer.alloc(32, 3)).bech32()

const mockSigner = {
  getAddress: vi.fn(async () => alice),
  signTransaction: vi.fn(async (_tx) => 'mock_sig_hex'),
} as unknown as MultiversXSigner

describe('ExactMultiversXScheme', () => {
  const scheme = new ExactMultiversXScheme(mockSigner)

  it('should create a valid payment payload', async () => {
    vi.mock('@multiversx/sdk-network-providers', () => ({
      ApiNetworkProvider: vi.fn().mockImplementation(() => ({
        getAccount: vi.fn().mockResolvedValue({ nonce: 7 }),
      })),
    }))

    const req: PaymentRequirements = {
      scheme: 'exact',
      payTo: bob,
      amount: '1000',
      asset: 'EGLD',
      network: 'multiversx:D',
      maxTimeoutSeconds: 300,
      extra: {
        resourceId: 'res-1',
        relayer: relayer,
      },
    }

    const { x402Version, payload } = await scheme.createPaymentPayload(1, req)
    const exactPayload = payload as ExactMultiversXPayload

    expect(x402Version).toBe(1)
    expect(exactPayload.signature).toBe('mock_sig_hex')
    expect(exactPayload.nonce).toBe(7)
    expect(exactPayload.value).toBe('1000')
    expect(exactPayload.sender).toBe(alice)
    expect(exactPayload.receiver).toBe(bob)
    expect(exactPayload.chainID).toBe('D')
    expect(exactPayload.version).toBe(2)
    expect(exactPayload.gasPrice).toBe(1_000_000_000)
    expect(exactPayload.relayer).toBe(relayer)

    const now = Math.floor(Date.now() / 1000)
    expect(exactPayload.validBefore).toBeGreaterThan(now)
    expect(exactPayload.validAfter).toBeLessThanOrEqual(now)
  })

  it('should create a valid direct EGLD payload (version 1)', async () => {
    const req: PaymentRequirements = {
      scheme: 'exact',
      payTo: bob,
      amount: '100',
      asset: 'EGLD',
      network: 'multiversx:D',
      maxTimeoutSeconds: 0,
      extra: {
        assetTransferMethod: 'direct',
      },
    }

    const { payload } = await scheme.createPaymentPayload(1, req)
    const exactPayload = payload as ExactMultiversXPayload

    expect(exactPayload.version).toBe(1)
    expect(exactPayload.relayer).toBeUndefined()
    expect(exactPayload.receiver).toBe(bob)
    expect(exactPayload.value).toBe('100')
  })

  it('should create a valid EGLD payload with scFunction', async () => {
    const req: PaymentRequirements = {
      scheme: 'exact',
      payTo: bob,
      amount: '100',
      asset: 'EGLD',
      network: 'multiversx:D',
      maxTimeoutSeconds: 0,
      extra: {
        scFunction: 'buy',
        arguments: ['01', '02'],
        relayer: relayer,
      },
    }

    const { payload } = await scheme.createPaymentPayload(1, req)
    const exactPayload = payload as ExactMultiversXPayload

    expect(exactPayload.data).toBe('buy@01@02')
    expect(exactPayload.receiver).toBe(bob)
    expect(exactPayload.value).toBe('100')
    expect(exactPayload.gasLimit).toBe(10_313_500) // Base(50k) + Data(9*1500) + Multi(200k) + Relayed(50k) + SC(10M)
  })

  it('should create a valid ESDT payload', async () => {
    const req: PaymentRequirements = {
      scheme: 'exact',
      payTo: bob,
      amount: '100',
      asset: 'TEST-123456',
      network: 'multiversx:D',
      maxTimeoutSeconds: 0,
      extra: {
        relayer: relayer,
      },
    }

    const { payload } = await scheme.createPaymentPayload(1, req)
    const exactPayload = payload as ExactMultiversXPayload

    expect(exactPayload.receiver).toBe(alice)
    expect(exactPayload.value).toBe('0')
    expect(exactPayload.data).toContain('MultiESDTNFTTransfer')
    expect(exactPayload.data).toContain(new Address(bob).hex())
    expect(exactPayload.data).toContain(Buffer.from('TEST-123456', 'utf8').toString('hex'))
    expect(exactPayload.gasLimit).toBe(10_475_500)
  })

  it('should handle EGLD-000000 alias as ESDT', async () => {
    const req: PaymentRequirements = {
      scheme: 'exact',
      payTo: bob,
      amount: '100',
      asset: 'EGLD-000000',
      network: 'multiversx:D',
      maxTimeoutSeconds: 0,
      extra: {
        relayer: relayer,
      },
    }

    const { payload } = await scheme.createPaymentPayload(1, req)
    const exactPayload = payload as ExactMultiversXPayload

    expect(exactPayload.value).toBe('0')
    expect(exactPayload.data).toContain('MultiESDTNFTTransfer')
    expect(exactPayload.data).toContain(Buffer.from('EGLD-000000', 'utf8').toString('hex'))
  })
})

File: go/mechanisms/multiversx/exact/client/scheme_test.go
package client

import (
	"context"
	"strings"
	"testing"

	"github.com/multiversx/mx-chain-core-go/data/api"
	"github.com/multiversx/mx-chain-core-go/data/transaction"
	"github.com/multiversx/mx-sdk-go/core"
	"github.com/multiversx/mx-sdk-go/data"

	"github.com/coinbase/x402/go/mechanisms/multiversx"

	"github.com/coinbase/x402/go/types"
)

// MockSigner matches ClientMultiversXSigner
type MockSigner struct {
	addr string
}

func (m *MockSigner) Address() string {
	return m.addr
}
func (s *MockSigner) PrivateKey() []byte {
	// Valid 32-byte seed
	return []byte{
		0x41, 0x3f, 0x42, 0x57, 0x5f, 0x7f, 0x26, 0xfa,
		0xd3, 0x31, 0x7a, 0x77, 0x87, 0x71, 0x21, 0x2f,
		0xdb, 0x80, 0x24, 0x58, 0x50, 0x98, 0x1e, 0x48,
		0xb5, 0x8a, 0x4f, 0x25, 0xe3, 0x44, 0xe8, 0xf9,
	}
}

func (m *MockSigner) Sign(ctx context.Context, message []byte) ([]byte, error) {
	return []byte("signature"), nil
}

const (
	// Valid Bech32 Addresses for Testing (Alice/Bob Devnet)
	testPayTo  = "erd1spyavw0956vq68xj8y4tenjpq2wd5a9p2c6j8gsz7ztyrnpxrruqzu66jx" // Alice (using Bob's valid address)
	testSender = "erd1spyavw0956vq68xj8y4tenjpq2wd5a9p2c6j8gsz7ztyrnpxrruqzu66jx" // Bob
	testAsset  = "TEST-123456"
	testAmount = "1000000000000000000" // 1 EGLD
)

// MockProxy implements Proxy interface
type MockProxy struct {
	nonce uint64
	err   error
}

// GetAccount must match blockchain.Proxy interface
func (m *MockProxy) GetAccount(ctx context.Context, address core.AddressHandler) (*data.Account, error) {
	return &data.Account{
		Nonce: m.nonce,
	}, m.err
}

func (m *MockProxy) SendTransaction(ctx context.Context, tx *transaction.FrontendTransaction) (string, error) {
	return "txHash", nil
}

func (m *MockProxy) GetNetworkConfig(ctx context.Context) (*data.NetworkConfig, error) {
	return &data.NetworkConfig{
		MinGasLimit: 50000,
		MinGasPrice: 1000000000,
		ChainID:     "D",
	}, nil
}

// IsInterfaceNil required by Proxy interface
func (m *MockProxy) IsInterfaceNil() bool {
	return m == nil
}

// Helpers methods required by Proxy interface (stubs)
func (m *MockProxy) SendTransactions(ctx context.Context, txs []*transaction.FrontendTransaction) ([]string, error) {
	return []string{"txHash"}, nil
}
func (m *MockProxy) GetGuardianData(ctx context.Context, address core.AddressHandler) (*api.GuardianData, error) {
	return nil, nil // Not used
}
func (m *MockProxy) ExecuteVMQuery(ctx context.Context, vmRequest *data.VmValueRequest) (*data.VmValuesResponseData, error) {
	return nil, nil // Not used
}
func (m *MockProxy) FilterLogs(ctx context.Context, filter *core.FilterQuery) ([]*transaction.Events, error) {
	return nil, nil // Not used
}

func TestCreatePaymentPayload_EGLD(t *testing.T) {
	signer := &MockSigner{addr: testSender}
	mockProxy := &MockProxy{nonce: 15}
	scheme, _ := NewExactMultiversXScheme(signer, "multiversx:D", WithProxy(mockProxy))

	req := types.PaymentRequirements{
		PayTo:   testPayTo,
		Amount:  "100",
		Asset:   "EGLD",
		Network: "multiversx:D",
		Extra: map[string]interface{}{
			"relayer": testSender,
		},
	}

	payload, err := scheme.CreatePaymentPayload(context.Background(), req)
	if err != nil {
		t.Fatalf("Failed to create payload: %v", err)
	}

	// Verify structure
	rpPtr, err := multiversx.PayloadFromMap(payload.Payload)
	if err != nil {
		t.Fatalf("Failed to parse payload: %v", err)
	}
	rp := *rpPtr

	if rp.Receiver != testPayTo {
		t.Errorf("Wrong receiver: %s", rp.Receiver)
	}
	if rp.Value != "100" {
		t.Errorf("Wrong value: %s", rp.Value)
	}
	if rp.Data != "" {
		t.Errorf("Expected empty data for EGLD, got %s", rp.Data)
	}
	if rp.Nonce != 15 {
		t.Errorf("Wrong nonce: %d", rp.Nonce)
	}
}

func TestCreatePaymentPayload_EGLD_WithScFunction(t *testing.T) {
	signer := &MockSigner{addr: testSender}
	mockProxy := &MockProxy{nonce: 15}
	scheme, _ := NewExactMultiversXScheme(signer, "multiversx:D", WithProxy(mockProxy))

	req := types.PaymentRequirements{
		PayTo:   testPayTo,
		Amount:  "100",
		Asset:   "EGLD",
		Network: "multiversx:D",
		Extra: map[string]interface{}{
			"scFunction": "buy",
			"arguments":  []string{"01", "02"},
			"relayer":    testSender,
		},
	}

	payload, err := scheme.CreatePaymentPayload(context.Background(), req)
	if err != nil {
		t.Fatalf("Failed to create payload: %v", err)
	}

	rpPtr, err := multiversx.PayloadFromMap(payload.Payload)
	if err != nil {
		t.Fatalf("Failed to parse payload: %v", err)
	}
	rp := *rpPtr

	// Should be plain text for native: buy@01@02
	expectedData := "buy@01@02"
	if rp.Data != expectedData {
		t.Errorf("Wrong data: expected %s, got %s", expectedData, rp.Data)
	}
}

func TestCreatePaymentPayload_ESDT(t *testing.T) {
	signer := &MockSigner{addr: testSender}
	mockProxy := &MockProxy{nonce: 20}
	scheme, _ := NewExactMultiversXScheme(signer, "multiversx:D", WithProxy(mockProxy))

	req := types.PaymentRequirements{
		PayTo:   testPayTo,
		Amount:  "100",
		Asset:   testAsset,
		Network: "multiversx:D",
		Extra: map[string]interface{}{
			"relayer": testSender,
		},
	}

	payload, err := scheme.CreatePaymentPayload(context.Background(), req)
	if err != nil {
		t.Fatalf("Failed to create payload: %v", err)
	}

	rpPtr, err := multiversx.PayloadFromMap(payload.Payload)
	if err != nil {
		t.Fatalf("Failed to parse payload: %v", err)
	}
	rp := *rpPtr

	// ESDT check: Receiver should be Sender (Self-transfer)
	if rp.Receiver != testSender {
		t.Errorf("ESDT tx receiver should be sender, got %s", rp.Receiver)
	}
	if rp.Value != "0" {
		t.Errorf("ESDT tx value should be 0 EGLD, got %s", rp.Value)
	}
	// Check Nonce
	if rp.Nonce != 20 {
		t.Errorf("Wrong nonce: %d", rp.Nonce)
	}

	// Check Data field contains "MultiESDTNFTTransfer"
	if !strings.HasPrefix(rp.Data, "MultiESDTNFTTransfer") {
		t.Errorf("Data should start with MultiESDTNFTTransfer, got %s", rp.Data)
	}
}

func TestCreatePaymentPayload_ESDT_WithResourceID(t *testing.T) {
	signer := &MockSigner{addr: testSender}
	mockProxy := &MockProxy{nonce: 25}
	scheme, _ := NewExactMultiversXScheme(signer, "multiversx:D", WithProxy(mockProxy))

	req := types.PaymentRequirements{
		PayTo:   testPayTo,
		Amount:  "100",
		Asset:   testAsset,
		Network: "multiversx:D",
		Extra: map[string]interface{}{
			"scFunction": "inv_123",
			"relayer":    testSender,
		},
	}

	payload, err := scheme.CreatePaymentPayload(context.Background(), req)
	if err != nil {
		t.Fatalf("Failed to create payload: %v", err)
	}

	rpPtr, err := multiversx.PayloadFromMap(payload.Payload)
	if err != nil {
		t.Fatalf("Failed to parse payload: %v", err)
	}
	rp := *rpPtr

	// Check encoded resource ID "inv_123" -> hex "696e765f313233"
	// Should be at the end
	expectedRidHex := "696e765f313233"
	if !strings.HasSuffix(rp.Data, expectedRidHex) {
		t.Errorf("Data should end with scFunction hex %s, got %s", expectedRidHex, rp.Data)
	}
}

func TestCreatePaymentPayload_EGLD_Alias(t *testing.T) {
	signer := &MockSigner{addr: testSender}
	mockProxy := &MockProxy{nonce: 30}
	scheme, _ := NewExactMultiversXScheme(signer, "multiversx:D", WithProxy(mockProxy))

	req := types.PaymentRequirements{
		PayTo:   testPayTo,
		Amount:  "100",
		Asset:   "EGLD-000000", // Should be treated as EGLD if handled or ESDT token otherwise
		Network: "multiversx:D",
		Extra: map[string]interface{}{
			"relayer": testSender,
		},
	}

	payload, err := scheme.CreatePaymentPayload(context.Background(), req)
	if err != nil {
		t.Fatalf("Failed to create payload: %v", err)
	}

	rpPtr, err := multiversx.PayloadFromMap(payload.Payload)
	if err != nil {
		t.Fatalf("Failed to parse payload: %v", err)
	}
	rp := *rpPtr

	// Should be ESDT transfer (MultiESDTNFTTransfer)
	// Because we treat EGLD-000000 as a token identifier for MultiESDT.

	// Value should be 0 (Native EGLD not sent via Value field in MultiESDT usually, unless implied?)
	// Actually, if using EGLD-000000 in MultiESDT, the 'value' of tx is 0, and amount is in data.
	if rp.Value != "0" {
		t.Errorf("Value should be 0 for MultiESDT, got %s", rp.Value)
	}

	if !strings.HasPrefix(rp.Data, "MultiESDTNFTTransfer") {
		t.Errorf("Data should start with MultiESDTNFTTransfer, got %s", rp.Data)
	}

	// Check token hex for EGLD-000000
	// "EGLD-000000" -> 45474c442d303030303030
	tokenHex := "45474c442d303030303030"
	if !strings.Contains(rp.Data, tokenHex) {
		t.Errorf("Data should contain EGLD-000000 hex %s, got %s", tokenHex, rp.Data)
	}
}
